src/main/scala/chisel3/tester/TreadleBackend.scala

```diff
@@ -0,0 +1,212 @@
+// See LICENSE for license details.
+
+package chisel3.tester
+
+import chisel3._
+import java.util.concurrent.{Semaphore, SynchronousQueue, TimeUnit}
+
+import scala.collection.mutable
+import firrtl_interpreter._
+import treadle.{HasTreadleSuite, TreadleTester}
+
+class TreadleBackend[T <: Module](dut: T, tester: TreadleTester)
+    extends BackendInstance[T] with ThreadedBackend {
+  def getModule() = dut
+
+  /** Returns a Seq of (data reference, fully qualified element names) for the input.
+    * name is the name of data
+    */
+  protected def getDataNames(name: String, data: Data): Seq[(Data, String)] = Seq(data -> name) ++ (data match {
```

Comment:
I think all of this should be done.  I am hesitant to do that in a PR like this because it makes it harder to compare with the source that was cut-and-pasted.
I'd be inclined to submit a second PR with the style based changes.  This code is richly highlighted by IntelliJ style checker.

Reply:
Cut and paste code is a huge no. Could you just replace the firrterpreter implementation with the treadle implementation? If needed in the future, I'll refactor things out cleanly.